_posts/2017-01-21-Wonders-of-Static-Linking.md

```diff
@@ -0,0 +1,147 @@
+
+We are taught how to compile a program, yes very easy process, most of us think its just pressing the **Compile** button or pressing the **Compile and Execute** button for those who use Dev-C/C++ or just use `gcc` who use linux. Its a good defination for a layman, but for a programmer, yes for us, its not enough. For the coders(some basic coders), like us, there are 4 steps of the compilation these are
+
+* Preprocessing
+* Compilation
+* Assembling
+* Linking
+
+The preprocessing is easy for C/C++ languages, it just replace the pre-program directive, the lines starting with `#`, like `#include <stdio.h>` with the file which has the **prototypes** of the functions which will be used or replace the Macros with their expansions like `#define POSEIDON 7`, this line will replace the occurence of POSEIDON with 7, pretty basic huh.
+
+Then is compilation, which checks the possiblities of the syntax and sementic errors if possible and then will initilize the assembling phase of the process.
+
+Then is the assembling, this will convert the language, the high level language which is C to the Assembly language, a low level language, which is hardware dependent. The C code after the preprocessing and compilation will be converted to the assembly language of the given hardware, which will be the .obj file, the object file.
+
+Then is the linking phase, the whole program even from small "Hello World" program to the big Windows like Operating systems, at some phase requires linking. So, what is linking, we usually understand or got some idea of the other processes but getting linking can be difficult for begineers, but I bet it is not difficult, what it is, is very simple concept. Now take the example, you make the "Hello World" program, 
+{% highlight c %}
+#include <stdio.h>
+
+int main()
+{
+    printf("Hello World\n");
+    return 0;
+}
+{% endhighlight %}
+
+During the compilation of this program, first `#include <stdio.h>` will be replaced by the content of the stdio.h file, which contains the prototypes of the functions like `int printf(const char *, ...);` which we used now, and the implementation of the file will be complied, yes we will have the compiled file of the functions of the stdio.h which are then **linked** with this hello.c file, so linking put together the different pieces of the like different functions together in a single file so the piece of the code for the printf will be linked with this hello world program.
+
+So this was the basic of the compilation, I hope the reader will understand whats linking, so it will be cool or HOT if we could also link our files which we made. There are two types of linking static and dynamic in this post I will discuss the static in this post.
```

Comment:
> So this was the basic of the compilation,

The `the` before compilation seems wrong to me, maybe replace with `So this was the basic idea of compilation`'

> or HOT

Don't like this, but if you want, we can keep it.

>There are two types of linking static and dynamic in this post I will discuss the static in this post.
 
These are two different sentences. Put a full stop after dynamic and a comma after linking.